Applications/Spire/Source/Ui/ScrollBarStyle.cpp

```diff
@@ -0,0 +1,152 @@
+}
+
+int ScrollBarStyle::pixelMetric(PixelMetric metric,
+    const QStyleOption* option, const QWidget* widget) const {
+  if(metric == PM_ScrollBarExtent) {
+    if(auto scroll_bar = qobject_cast<const QScrollBar*>(widget); scroll_bar) {
+      if(scroll_bar->orientation() == Qt::Horizontal) {
+        return m_horizontal_handle_height;
+      } else {
+        return m_vertical_handle_width;
+      }
+    }
+  }
+  return QProxyStyle::pixelMetric(metric, option, widget);
+}
+
+int ScrollBarStyle::styleHint(QStyle::StyleHint hint,
+    const QStyleOption* option, const QWidget* widget,
+    QStyleHintReturn* return_data) const {
+  if(hint == QStyle::SH_ScrollBar_ContextMenu) {
+    return 0;
+  }
+  return QProxyStyle::styleHint(hint, option, widget, return_data);
+}
+
+QRect ScrollBarStyle::subControlRect(ComplexControl control,
+    const QStyleOptionComplex* option, SubControl sub_control,
+    const QWidget* widget) const {
+  if(control == CC_ScrollBar) {
+    if(auto scroll_bar = qobject_cast<const QScrollBar*>(widget); scroll_bar) {
+      switch(sub_control) {
+        case SC_ScrollBarAddPage:
+          if(scroll_bar->orientation() == Qt::Horizontal) {
+            auto slider_right_pos = get_horizontal_slider_position(
+              scroll_bar) + get_handle_size(scroll_bar);
+            return {slider_right_pos, 0,
+              scroll_bar->width() - slider_right_pos,
+              m_horizontal_handle_height};
+          } else {
+            auto slider_bottom_pos = get_vertical_slider_position(
+              scroll_bar) + get_handle_size(scroll_bar);
+            return {0, slider_bottom_pos, m_vertical_handle_width,
+              scroll_bar->height() - slider_bottom_pos};
+          }
+        case SC_ScrollBarSubPage:
+          if(scroll_bar->orientation() == Qt::Horizontal) {
+            return {0, 0, get_horizontal_slider_position(scroll_bar),
+              m_horizontal_handle_height};
+          }
+          return {0, 0, m_vertical_handle_width,
+            get_vertical_slider_position(scroll_bar)};
+        case SC_ScrollBarGroove:
+          return widget->rect();
+        case SC_ScrollBarSlider:
+          if(scroll_bar->orientation() == Qt::Horizontal) {
+            return {get_horizontal_slider_position(scroll_bar), 0,
+              get_handle_size(scroll_bar), scroll_bar->height()};
+          }
+          return {0, get_vertical_slider_position(scroll_bar),
+            scroll_bar->width(), get_handle_size(scroll_bar)};
+        case SC_ScrollBarAddLine:
+        case SC_ScrollBarFirst:
+        case SC_ScrollBarLast:
+        case SC_ScrollBarSubLine:
+          return {};
+      }
+    }
+  }
+  return QProxyStyle::subControlRect(control, option, sub_control, widget);
+}
+
+void ScrollBarStyle::set_horizontal_scroll_bar_height(int height) {
+  m_horizontal_handle_height = height;
+}
+
+void ScrollBarStyle::set_vertical_scroll_bar_width(int width) {
+  m_vertical_handle_width = width;
```

Comment:
Does an update need to be performed? I'm concerned that a state is being changed here but no associated side effect is produced. This could be fine but it looks fishy.

Reply:
It's possible that we could set a style on an existing scroll bar (like a QListWidget's scroll bar) and we would have to call update() in that case.

In practice only the ScrollArea calls these methods so this isn't an issue because the ScrollArea is calling update() after calling these, but, to keep it working predictably it's probably best to have the style call update on the given parent QWidget.

Reply:
No it's fine then, keep it as is.